### ISSUE TITLE
[Bug] moreh_softmin returns all-zero rows for inputs containing ±inf on every kernel path; the SOFTMIN compute define is never consumed by any kernel

### ISSUE THREAD

**Issue by heroxcreed** [tenstorrent/tt-metal#56371]
## Summary

`ttnn.operations.moreh.softmin` silently returns **all-zero rows** whenever the reduced dimension contains `+inf` (torch `F.softmin` returns a well-defined distribution with 0 at the +inf position), and all-zero rows for `-inf` inputs as well (torch returns an all-NaN row). The defect reproduces on **every** parallelization path (W small, W large, H, NC, dim 0/1, masked/unaligned W). No error is raised — the output is silently wrong.

Reproduced end-to-end on the Blackhole ttsim simulator (no hardware needed).

## Reproduction

```python
import torch, ttnn
import torch.nn.functional as F

device = ttnn.open_device(device_id=0)

x = torch.randn(1, 1, 32, 32)
x[0, 0, 0, 5] = float("inf")

tt = ttnn.from_torch(x.bfloat16(), dtype=ttnn.bfloat16, layout=ttnn.TILE_LAYOUT, device=device)
out = ttnn.operations.moreh.softmin(tt, 3)
got = ttnn.to_torch(out).float()

ref = F.softmin(x.float(), 3)
print(got[0,0,0,:8])   # all zeros
print(ref[0,0,0,:8])   # [0.0090, 0.0197, 0.0086, 0.0056, 0.0280, 0.0, 0.0091, 0.0278]
print(got.sum(dim=3)[0,0,0].item())  # 0.0
print(ref.sum(dim=3)[0,0,0].item())  # 1.0
```

## Evidence matrix (ttsim, Blackhole, ttnn 0.78.0 wheel semantics)

| case | torch `F.softmin` | ttnn `moreh.softmin` |
|---|---|---|
| dim=3, W=32 (single tile), one +inf in row | row sums to 1.0, 0 at +inf pos | **entire row 0** |
| dim=3, W=128 (4 tiles, large-W path), one +inf | row sums to 1.0 | **entire row 0** |
| dim=2, H=128 (H path), one +inf | col sums to 1.0 | **entire col 0** |
| dim=1 / dim=0 (NC path), one +inf | sums to 1.0 | wrong distribution (max err 0.36/0.55) |
| dim=3, W=33 unaligned (mask path), one +inf | row sums to 1.0 | **entire row 0** |
| one −inf in row | all-NaN row (torch) | **entire row 0** |
| all−inf / all+inf row | sum 0 / NaN | sum ≈ 1 (wrong) |
| finite inputs | — | correct (max err ≈0.005, bf16 noise) |

Sibling ops are **not** affected: `moreh.softmax` and `moreh.logsoftmax` handle a +inf row correctly (verified on the same setup: row = [0,…,1 at inf pos,…,0], matching torch).

## Root cause

The program factories define `SOFTMIN=1` for the softmin op
(`ttnn/cpp/ttnn/operations/moreh/moreh_softmax/device/softmax_w_small/softmax_w_small.cpp:167`
and the h/w/c factory variants), but **no kernel in
`ttnn/cpp/ttnn/operations/moreh/moreh_softmax/device/kernels/` contains an
`#ifdef SOFTMIN` / `#ifndef SOFTMIN` branch** — the define is dead. All kernels implement:

```
m   = max(x)            // row/col max of the ORIGINAL input
e   = −(x − m)          // #ifndef SOFTMAX → negative_tile after the subtraction
out = exp(e) / Σ exp(e)
```

For **finite** inputs this equals `softmax(−x)` = `softmin(x)` by the algebraic identity
`exp(max(x)−x)/Σexp(max(x)−x) ≡ exp(min(x)−x)/Σexp(min(x)−x)` (numerator and denominator
both scaled by `exp(max−min)`), which is why nightly tests pass. For **non-finite** inputs
the identity breaks: with `+inf` in the row, `m = inf`, every finite element gets
`exp(inf−x) = inf`, `Σ = inf`, `1/Σ = 0`, and the whole row collapses to 0. The stable
statistic used for subtraction must be the max over the **negated** input (`max(−x) = −min(x)`),
the way `moreh_norm` already special-cases `p = −inf`.

## Proposed fix direction

Mirror the existing, already-shipped pattern from `moreh_norm`'s `MINUS_INF` order handling
(`moreh_norm_program_factory_{h,w,nc}_other.cpp` + `moreh_norm_{h,w,nc}_kernel.cpp`):
for the softmin variant, compute the row statistic as the max over the negated input (or
equivalently `min(x)`), so the subtracted quantity is finite whenever any element is
finite, and `exp()` saturates to 0 at ±inf positions exactly like torch. An `#ifdef SOFTMIN`
branch in the five softmax kernels (`moreh_softmax_{h,w}.cpp`, `moreh_softmax_{h,w}_large.cpp`,
`moreh_softmax_c_large.cpp`) around the max-reduce/subtraction makes the currently-dead
define meaningful again.

Known limitation to document: for a row containing `−inf` (or all-±inf rows), torch
produces NaN via IEEE `inf−inf`; the SFPU FPU evaluates `inf−inf` to 0, so a fixed kernel
will produce a well-defined distribution concentrating on the minimum instead — same
class of divergence already documented for the p=±inf norm orders (#56248, #56334).

## Test-coverage gap

`tests/ttnn/nightly/unit_tests/operations/moreh/test_moreh_softmin.py` parameterizes
only finite inputs (`torch.rand(...) + 100` / `randint(0,4)+100`); no case feeds ±inf or
NaN into any dim. The same gap existed for `moreh_norm_backward`'s p parameterization
(p ∈ [2.0, 2.5, −2.5] only, ±inf never covered — #56248).

**Comment by heroxcreed** [comment]
I'd like to take this one.

Plan:
1. Fix: add an `#ifdef SOFTMIN` branch in the five moreh_softmax kernels (`moreh_softmax_{h,w}.cpp`, `moreh_softmax_{h,w}_large.cpp`, `moreh_softmax_c_large.cpp`) so the row/col statistic is computed over the negated input (max(−x) = −min(x)), mirroring the already-shipped `MINUS_INF` pattern from `moreh_norm`'s ord_other kernels — the `SOFTMIN` define the factories already emit then becomes meaningful.
2. Verification: before/after on the Blackhole ttsim simulator against `torch.nn.functional.softmin` for the full path matrix in the issue (W small/large, H, NC dim 0/1, masked/unaligned W, ±inf single-element / all-±inf rows, finite baseline).
3. Regression test: parameterize `test_moreh_softmin.py` with ±inf cases under the same shapes nightly already covers.

Please assign.

**Comment by heroxcreed** [comment]
Fix is up: #56386 (7 files, +321/−5).

**Approach** — the `SOFTMIN` define the factories already emit is now consumed by the five forward kernels: the row/column statistic for softmin becomes `min(x)` (computed as `-max(-x)`, since the FPU has no bf16 MIN reduce; the `c_large` kernel folds with `binary_min_tile` directly). Masked padding lanes are written as `+inf` (`mask_posinf_tile`) before the negate so they become `-inf` and can never win the `max(-x)` reduce — mirroring the `MINUS_INF` pattern already shipped in the `moreh_norm` ord_other kernels. The downstream `x − m` / exp / normalize stages are unchanged; the softmax paths are untouched behind `#ifdef SOFTMAX`.

**Verified on the Blackhole ttsim simulator** (no hardware): the full before/after matrix from this issue across every kernel path (W small/large, H, NC dim 0/1, unaligned/masked) — +inf rows now match torch (sum ≈ 1, max err ≤ 0.005 vs previous all-zero), NC-path max err 0.36/0.55 → 0.024/0.022. Full nightly `test_moreh_softmin.py` 104 passed / 32 skipped (bfloat8_b, official skip) / 0 failed, including 12 new `test_softmin_special_value_in_row` cases; the shared-kernel `test_moreh_softmax.py` + `test_moreh_logsoftmax.py` suites pass 185/0 as well.

The −inf / all-±inf divergence documented in the issue (SFPU `inf−inf` = 0 vs torch NaN) is asserted as documented behavior in the regression test.

**Comment by heroxcreed** [comment]
Hi @jberkowitzTT — I reported this and would like to take it. The fix is already up as PR #56386 (7 files, +321/-5), with the reproduction and root-cause analysis in the issue body.
Could you assign the issue to me? I'm happy to work through review feedback.
